### PR TITLE
Ensure order of operations for HTTPClient

### LINKED ISSUE #750: Flaky test PurchasesTests/Networking/HTTPClientTests.swift:292
https://app.circleci.com/pipelines/github/RevenueCat/purchases-ios/3006/workflows/5768965a-3626-46fd-8406-64c3173ff343/jobs/5472/tests#failed-test-0
`testParseError() `

### PR DESCRIPTION
**Pull request by taquitos**
There were a few places where we were queue-hopping a bit. The original implementation locked in just a few places. With the Swift migration we introduced a couple changes, like always dispatching `completion` to the main queue and wrapping the `handler` and `perform` methods in a dispatch work item. The data task we create dispatches the completion to the Session's own internal queue (we don't have a delegate), so we end up potentially using 4 queues here, original caller, serial http, internal session, then main. I reduced the queue usage to just two, the original caller and the internal queue on Session. I also used NSRecursiveLock for access control (to replace the original `@synchronized`).

resolves #750

### CODE REVIEW

**Comment by aboedo** [comment]
since it's WIP I'm updating the status to draft

File: Purchases/Networking/HTTPClient.swift [hunk 1]
```diff
@@ -21,6 +21,7 @@ class HTTPClient {
     private var currentSerialRequest: HTTPRequest?
     private var eTagManager: ETagManager
     private let operationDispatcher: OperationDispatcher
```
Comment (aboedo) [inline Purchases/Networking/HTTPClient.swift:23]: is `operationDispatcher` still used in this class? if not, can we remove it?

File: PurchasesTests/Purchasing/PurchasesTests.swift [hunk 2]
```diff
@@ -2354,7 +2354,7 @@ class PurchasesTests: XCTestCase {
         setupPurchases()
 
         for key in data.keys {
-            expect(self.backend.invokedPostAttributionDataParameters?.data?.keys.contains(key)) == true
+            expect(self.backend.invokedPostAttributionDataParameters?.data?.keys.contains(key)).toEventually(beTrue())
```
Comment (aboedo) [inline PurchasesTests/Purchasing/PurchasesTests.swift:2357]: we don't really need this `toEventually` to happen on all of them, right? 
perhaps we can instead have this once before the `for` loop
```swift
expect(self.backend.invokedPostAttributionData).toEventually(beTrue())
```
Which also indicates to readers that we're waiting for the method to have been invoked, before doing all of the checks. 
This also matches our expectation that once once parameter was set, they all have, we're not waiting until the next one

**Review by aboedo — APPROVED** [review]
looks great! thanks for taking care of this!

File: Purchases/Networking/HTTPClient.swift (hunk 1 above)
Comment (taquitos) [reply Purchases/Networking/HTTPClient.swift:23]: Oh hey, look at that. Yeah!

Comment (taquitos) [reply Purchases/Networking/HTTPClient.swift]: ✅

File: PurchasesTests/Purchasing/PurchasesTests.swift (hunk 2 above)
Comment (taquitos) [reply PurchasesTests/Purchasing/PurchasesTests.swift:2356]: Yeah, I was going to say it doesn't matter, BUT I really like your point about indicating to the reader that we're waiting for the method to be invoked and expect the rest to be there at that point. Updated!

Comment (taquitos) [reply PurchasesTests/Purchasing/PurchasesTests.swift]: ✅

**Comment by taquitos** [comment]
> looks great! thanks for taking care of this!

It takes teamwork to make the dream work 

spongebob-rainbow

**Pull request merged**